command/operator_raft_snapshot_inspect.go

```diff
@@ -0,0 +1,667 @@
+	if c.details {
+		if val.Key == "" {
+			return
+		}
+
+		// check for whether a filter is specified. if it is, skip
+		// any keys that don't match.
+		if len(c.filter) > 0 && !strings.HasPrefix(val.Key, c.filter) {
+			return
+		}
+
+		split := strings.Split(string(val.Key), "/")
+
+		// handle the situation where the key is shorter than
+		// the specified depth.
+		actualDepth := c.depth
+		if c.depth == 0 {
+			actualDepth = len(split)
+		} else if c.depth > len(split) {
+			actualDepth = len(split)
+		}
+
+		prefix := strings.Join(split[0:actualDepth], "/")
+		kvs := info.StatsKV[prefix]
+		if kvs.Name == "" {
+			kvs.Name = prefix
+		}
+
+		kvs.Count++
+		kvs.Size += read
+		info.TotalCountKV++
+		info.TotalSizeKV += read
+		info.StatsKV[prefix] = kvs
+	}
+}
+
+// Read from snapshot's state.bin and update the SnapshotInfo struct
+func (c *OperatorRaftSnapshotInspectCommand) parseState(r io.Reader) (SnapshotInfo, error) {
+	info := SnapshotInfo{
+		StatsKV:      make(map[string]typeStats),
+		TotalCountKV: 0,
+		TotalSizeKV:  0,
+	}
+	handler := func(s *pb.StorageEntry, read int) error {
+		c.kvEnhance(s, &info, read)
+		return nil
+	}
+	protoReader := protoio.NewDelimitedReader(r, math.MaxInt32)
+	errCh := make(chan error, 1)
+
+	go func() {
+		for {
+			s := new(pb.StorageEntry)
+			if err := protoReader.ReadMsg(s); err != nil {
+				if err == io.EOF {
+					errCh <- nil
+					return
+				}
+				errCh <- err
+				return
+			}
+			size := protoReader.GetLastReadSize()
+			handler(s, size)
+		}
+	}()
+
+	err := <-errCh
+	if err != nil && err != io.EOF {
+		return info, err
+	}
+
+	return info, nil
+}
+
+// ReadSnapshot decodes each message type and utilizes the handler function to
+// process each message type individually
+func ReadSnapshot(r io.Reader, handler func(s *pb.StorageEntry, read int) error) (*iradix.Tree, error) {
+	reader := protoio.NewDelimitedReader(r, math.MaxInt32)
+
+	errCh := make(chan error, 1)
+	txn := iradix.New().Txn()
+
+	go func() {
+		for {
+			s := new(pb.StorageEntry)
+
+			err := reader.ReadMsg(s)
+			if err != nil {
+				if err == io.EOF {
+					errCh <- nil
+					return
+				}
+				errCh <- err
+				return
+			}
+
+			size := reader.GetLastReadSize()
+
+			handler(s, size)
+
+			var value interface{} = struct{}{}
+			value = s.Value
+
+			txn.Insert([]byte(s.Key), value)
+		}
+	}()
+
+	err := <-errCh
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return txn.Commit(), nil
+}
+
+// Read contents of snapshot. Parse metadata and snapshot info
+// Also, verify validity of snapshot
+func (c *OperatorRaftSnapshotInspectCommand) Read(logger hclog.Logger, in io.Reader) (*SnapshotInfo, *raft.SnapshotMeta, error) {
+	// Wrap the reader in a gzip decompressor.
+	decomp, err := gzip.NewReader(in)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decompress snapshot: %v", err)
+	}
+	defer func() {
+		if err := decomp.Close(); err != nil {
```

Comment:
Rather than log this error, why not return it?

Reply:
Would that work in a defer?

Reply:
It can be made to work, sure.  At the end of your func:

```
if err := decomp.Close(); err != nil {
  return nil, nil, err
}
decomp = nil
return snapshotInfo, metadata, nil
```

Then in your defer:

```
defer func() {
  if decomp != nil {
    decomp.Close()
  }
}()
```

We don't care about an error closing the read handle if we already encountered an error earlier.

Reply:
ah I see, thank you! I added the log in the defer just in case, it fails somewhere before the `decomp.Close` here https://github.com/hashicorp/vault/pull/23457/files#diff-a95de5b9f984efafb6e2d327a0ade2e7e352526647cef32560988e5ad0528e3eR302